awx/main/access.py

```diff
@@ -2702,6 +2703,17 @@ class RoleAccess(BaseAccess):
     '''
 
     model = Role
+    prefetch_related = ('content_type',)
+
+    def filtered_queryset(self):
+        result = Role.visible_roles(self.user)
+        # Sanity check: is the requesting user an orphaned non-admin/auditor?
+        # if yes, make system admin/auditor mandatorily visible.
+        if not self.user.organizations.exists() and not self.user.is_superuser and not self.user.is_system_auditor:
+            mandatories = ('system_administrator', 'system_auditor')
+            super_qs = Role.objects.filter(singleton_name__in=mandatories)
+            result = result | super_qs
+        return result
```

Comment:
this is the riskiest part of this change. Tested for `/api/v2/roles/`, naturally, but the sublists are still awaiting confirmation.

Reply:
nevermind this comment, the object sublists define `get_queryset`, unfiltered to the current user. I don't know if any role sublists starts from this queryset.